src/libraries/System.Private.Uri/src/System/IriHelper.cs

```diff
@@ -68,114 +94,57 @@ internal static bool CheckIsReserved(char ch, UriComponents component)
         internal static unsafe string EscapeUnescapeIri(char* pInput, int start, int end, UriComponents component)
         {
             int size = end - start;
-            ValueStringBuilder dest = new ValueStringBuilder(size);
-            byte[]? bytes = null;
-
-            int next = start;
-            char ch;
+            ValueStringBuilder dest = size <= 256
+                ? new ValueStringBuilder(stackalloc char[256])
+                : new ValueStringBuilder(size);
 
             Span<byte> maxUtf8EncodedSpan = stackalloc byte[4];
 
-            for (; next < end; ++next)
+            for (int i = start; i < end; ++i)
             {
-                if ((ch = pInput[next]) == '%')
+                char ch = pInput[i];
+                if (ch == '%')
                 {
-                    if (next + 2 < end)
+                    if (i + 2 < end)
                     {
-                        ch = UriHelper.EscapedAscii(pInput[next + 1], pInput[next + 2]);
+                        ch = UriHelper.EscapedAscii(pInput[i + 1], pInput[i + 2]);
 
                         // Do not unescape a reserved char
                         if (ch == Uri.c_DummyChar || ch == '%' || CheckIsReserved(ch, component) || UriHelper.IsNotSafeForUnescape(ch))
                         {
                             // keep as is
-                            dest.Append(pInput[next++]);
-                            dest.Append(pInput[next++]);
-                            dest.Append(pInput[next]);
+                            dest.Append(pInput[i++]);
+                            dest.Append(pInput[i++]);
+                            dest.Append(pInput[i]);
```

Comment:
Each `Append` will perform a bounds check.  Can you use the `Append(char *, int)` overload here?

Reply:
That overload will do a Span slice as well. I was thinking of adding `Append(char, char)` and `Append(char, char, char)` overloads as they can have a measurable perf impact (that would be part of a separate PR adressing #22903).

Reply:
`Append(char, char)` and `Append(char, char, char)` seem quite awkward API choices, IMHO.  Maybe make the `Append(char *, int)` overload not create a Span slice?